src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java

```diff
@@ -104,35 +111,90 @@ public List<LocalCIBuildJobQueueItem> getProcessingJobsForCourse(long courseId)
         return processingJobs.values().stream().filter(job -> job.getCourseId() == courseId).toList();
     }
 
-    /**
-     * Get first build job item from the queue. If it exists, process build job and after completion,
-     * try to process next item.
-     */
-    private void processBuild() {
+    private ProgrammingExerciseParticipation retrieveParticipationWithRetry(Long participationId) {
+        int maxRetries = 5;
+        int retries = 0;
+        while (retries < maxRetries) {
+            try {
+                return (ProgrammingExerciseParticipation) participationRepository.findByIdElseThrow(participationId);
```

Comment:
Is there a method `findById()` that returns an `Optional`? I think that makes more sense than exception-based control flow here.